File: CCC_2022/CCC_2022_Problem_S5_Good_Influencers.java
import java.util.*;
public class CCC_2022_Problem_S5_Good_Influencers {
	static int MAX = (int) (2e5+5);
	static int[][][] dp = new int [MAX][2][2];
	static int[][] tmpdp = new int [2][2];
	static int[] cost = new int [MAX]; 
	static LinkedList<Integer> graph[] = new LinkedList[MAX];
	static boolean[] want = new boolean[MAX];
	public static void main(String[] args) {
		Scanner sc = new Scanner(System.in);
		int N = sc.nextInt();
		for (int i=0; i<MAX; i++){
            graph[i] = new LinkedList<>();
        }
	    for(int i=1, a, b; i<N; i++){
	        a = sc.nextInt(); b = sc.nextInt();
	        graph[a].add(b);
	        graph[b].add(a);
	    }
	    String s = sc.next();
	    for(int i=1; i<=N; i++){
	        want[i] = (s.charAt(i-1) == 'Y');
	        cost[i] = sc.nextInt();
	    }
	    dfs(1, 0);
	    System.out.println(Math.min(dp[1][1][0], dp[1][1][1]));
	    sc.close();
	}
	static void dfs(int cur, int pre){
	    dp[cur][0][0] = 0;
	    dp[cur][0][1] = cost[cur];
	    dp[cur][1][0] = want[cur]?0:0x3f3f3f3f;
	    dp[cur][1][1] = want[cur]?cost[cur]:0x3f3f3f3f;
	    for(int nxt: graph[cur]){
	        if(nxt == pre) continue;
	        dfs(nxt, cur);
	        for(int i=0; i<2; i++) {
	        	for(int j=0; j<2; j++) {
	        		tmpdp[i][j] = 0x3f3f3f3f;
	        	}
	        }
	        for(int i=0; i<2; i++){
	            for(int j=0; j<2; j++){
	                for(int m=0; m<2; m++){
	                    for(int n=0; n<2; n++){
	                        if(j==0 && m==0) continue;
	                        int p = i|(m&n);
	                        int q = j;
	                        tmpdp[p][q] = Math.min(tmpdp[p][q], dp[cur][i][j] + dp[nxt][m][n]);
	                    }
	                }
	            }
	        }
	        for(int i=0; i<2; i++) {
	        	for(int j=0; j<2; j++) {
	        		dp[cur][i][j] = tmpdp[i][j];
	        	}
	        }
	    }
	}
}

File: CCC_2022/CCC_2022_Problem_S5_Good_Influencers.py
import sys
sys.setrecursionlimit(100000000)
graph = [[] for i in range(200005)]
want = [0]*200005
dp = [[[0 for k in range(2)] for j in range(2)] for i in range(200005)]
tmpdp = [[0 for j in range(2)] for i in range(2)]
N = int(input())
for i in range(N-1):
    a, b = map(int, input().split())
    graph[a].append(b)
    graph[b].append(a)
s = input()
for i in range(N):
    want[i+1] = (s[i] == 'Y')
cost = [0]+list(map(int, input().split()))

def dfs(cur, pre):
    dp[cur][0][0] = 0
    dp[cur][0][1] = cost[cur]
    if want[cur]:
        dp[cur][1][0] = 0
        dp[cur][1][1] = cost[cur]
    else:
        dp[cur][1][0] = dp[cur][1][1] = 123456789101112
    for nxt in graph[cur]:
        if nxt == pre:
            continue
        dfs(nxt, cur)
        for i in range(2):
            for j in range(2):
                tmpdp[i][j] = 123456789101112
        for i in range(2):
            for j in range(2):
                for m in range(2):
                    for n in range(2):
                        if not j and not m:
                            continue
                        p = i|(m&n)
                        q = j
                        tmpdp[p][q] = min(tmpdp[p][q], dp[cur][i][j] + dp[nxt][m][n]);
        for i in range(2):
            for j in range(2):
                dp[cur][i][j] = tmpdp[i][j]
dfs(1, 0)
print(min(dp[1][1][0], dp[1][1][1]))
